File: os-sim/www/av_center/data/sections/home/network.php
<?php

/**
 *
 * License:
 *
 * Copyright (c) 2003-2006 ossim.net
 * Copyright (c) 2007-2014 AlienVault
 * All rights reserved.
 *
 * This package is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 dated June, 1991.
 * You may not use, modify or distribute this program under any other version
 * of the GNU General Public License.
 *
 * This package is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this package; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 * MA  02110-1301  USA
 *
 *
 * On Debian GNU/Linux systems, the complete text of the GNU General
 * Public License can be found in `/usr/share/common-licenses/GPL-2'.
 *
 * Otherwise you can read it here: http://www.gnu.org/licenses/gpl-2.0.txt
 *
 */


//Config File
require_once dirname(__FILE__).'/../../../config.inc';

session_write_close();

if ($_SERVER['SCRIPT_NAME'] != '/ossim/av_center/data/sections/home/network.php')
{
    exit();
}

$system_id     = POST('system_id');
$force_request = (1 == POST('force_request')) ? TRUE : FALSE;

ossim_valid($system_id, OSS_DIGIT, OSS_LETTER, '-', 'illegal:'._('System ID'));

$error_msg = NULL;

if (ossim_error())
{
    $error_msg = _('System ID not found. Information not available');

    echo 'error###'.$error_msg;
    exit();
}

try
{
    $st = Av_center::get_system_status($system_id, 'network', $force_request);
}
catch (\Exception $e)
{
    echo 'error###'.$e->getMessage();
    exit();
}


/*************************************************************
***********************  Network Data  ***********************
*************************************************************/

$dns_servers = $st['dns_servers'];
$dns         = (is_array($dns_servers) && !empty($dns_servers)) ? 'tick.png' : 'cross.png';

$firewall_active  = ('yes' === $st['firewall_active'])     ? 'tick.png' : 'cross.png';
$internet         = ('yes' === $st['internet_connection']) ? 'tick.png' : 'cross.png';
$vpn_access       = ('yes' === $st['vpn_access'])          ? 'tick.png' : 'cross.png';


$img_firewall = AVC_PIXMAPS_DIR.'/'.$firewall_active;
$img_internet = AVC_PIXMAPS_DIR.'/'.$internet;
$img_vpn      = AVC_PIXMAPS_DIR.'/'.$vpn_access;
$img_dns      = AVC_PIXMAPS_DIR.'/'.$dns;
$img_rx       = AVC_PIXMAPS_DIR.'/'.'arrow-small-down-green.png';
$img_tx       = AVC_PIXMAPS_DIR.'/'.'arrow-small-up-red.png';

?>

<table id='t_network' cellspacing='0' cellpadding='0'>
    <tr>
        <th class='th_network' colspan='10'><?php echo _('General Information')?></th>
    </tr>

    <tr>
        <td class='_label' valign='middle'><label><?php echo _('Firewall')?></label></td>
        <td class='_data'><img id='firewall' src='<?php echo $img_firewall?>' alt='<?php echo $firewall_active?>' align='absmiddle'/></td>
        <td class='_label' valign='middle'><label><?php echo _('VPN Infrastructure')?></label></td>
        <td class='_data'><img id='vpn_access' src='<?php echo $img_vpn?>' alt='<?php echo $vpn_infrastructure?>' align='absmiddle'/></td>
        <td class='_label' valign='middle'><label><?php echo _('Internet Connection')?></label></td>
        <td class='_data'><img id='inet_conn' src='<?php echo $img_internet?>' alt='<?php echo $internet?>' align='absmiddle'/></td>
        <td class='_label' valign='middle'><label><?php echo _('Default Gateway')?></label></td>
        <td class='_data'><div id='gateway'><?php echo $st['gateway']?></div></td>
        <td class='_label' valign='middle'><label><?php echo _('DNS Servers')?></label></td>
        <td class='_data'>
        <?php
        if (empty($dns_servers))
        {
            ?>
            <div id='dns_servers'><img src='<?php echo $img_dns?>' alt='<?php echo $dns?>' align='absmiddle'/></div>
            <?php
        }
        else
        {
            ?>
            <div id='dns_servers'><div><?php echo str_replace(',', '</div><div>', $dns_servers)?></div></div>
            <?php
        }
        ?>
        </td>
    </tr>

    <tr>
        <th class='th_network' colspan='10'><?php echo _('Interface Information')?></th>
    </tr>

    <?php
    $i_role_names = array(
        'disabled'       => _('Not in Use'),
        'admin'          => _('Management'),
        'monitoring'     => _('Network Monitoring'),
        'log_management' => _('Log Collection & Scanning')
    );


    //Reorder interfaces
    $i_lo = array('lo' => $st['interfaces']['lo']);

    unset($st['interfaces']['lo']);
    $st['interfaces'] = array_merge($i_lo, $st['interfaces']);

    foreach($st['interfaces'] as $i_name => $i_data)
    {                
        if ('up' === $i_data['status'])
        {
             $text_color = 'green';
             $if_image   = 'port_animado.gif';
        }
        else
        {
             $text_color = 'red';
             $if_image   = 'no_animado.gif';
        }

        $i_data['rx_bytes'] = Avc_utilities::bytesToSize($i_data['rx_bytes'], 2);
        $i_data['tx_bytes'] = Avc_utilities::bytesToSize($i_data['tx_bytes'], 2);
        $i_data['role']     = ($i_name == 'lo') ? ' - ' : $i_role_names[$i_data['role']];
        ?>

        <tr>
            <td rowspan='2' class='_label td_iface_name'><?php echo $i_name?></td>
            <td rowspan='2' id='<?php echo $i_name.'_status'?>' class='td_iface_status'>
                <div id='container_iface_st'>
                    <img src="<?php echo AVC_PIXMAPS_DIR.'/'.$if_image?>" width="40"/>
                    <div class='iface_status'><span class='<?php echo $text_color?>'><?php echo strtoupper($i_data['status'])?></span></div>
                </div>
            </td>
            <td class='_label'><img style='margin-right: 3px;' src='<?php echo $img_rx?>' alt='&#8595;'/>Rx</td>
            <td class='_data' id='<?php echo $i_name.'_rx_bytes'?>'><?php echo $i_data['rx_bytes']?></td>
            <td class='_label'>IP</td>
            <td class='_data' colspan="2" id='<?php echo $i_name.'_address'?>'><?php echo $i_data['ipv4']['address']?></td>
            <td class='_label'><?php echo _('Role')?></td>
            <td class='_data' colspan="2"><?php echo $i_data['role']?></td>
        </tr>
        <tr>
            <td class='_label nobborder'><img style='margin-right: 3px;' src='<?php echo $img_tx?>' alt='&#8593;'/>Tx</td>
            <td class='_data' id='<?php echo $i_name.'_tx_bytes'?>'><?php echo $i_data['tx_bytes']?></td>
            <td class='_label'><?php echo _('Netmask')?></td>
            <td class='_data' colspan="2" id='<?php echo $i_name.'_netmask'?>'><?php echo $i_data['ipv4']['netmask']?></td>
            <td class='_label nobborder'><?php echo _('Network')?></td>
            <td class='_data nobborder' colspan="2" id='<?php echo $i_name.'_network'?>'><?php echo $i_data['ipv4']['network']?></td>
        </tr>
        <?php
    }
    ?>
</table>

File: alienvault-api/alienvault-api/src/blueprints/system/network.py
#
#  License:
#
#  Copyright (c) 2013 AlienVault
#  All rights reserved.
#
#  This package is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; version 2 dated June, 1991.
#  You may not use, modify or distribute this program under any other version
#  of the GNU General Public License.
#
#  This package is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this package; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
#  MA  02110-1301  USA
#
#
#  On Debian GNU/Linux systems, the complete text of the GNU General
#  Public License can be found in `/usr/share/common-licenses/GPL-2'.
#
#  Otherwise you can read it here: http://www.gnu.org/licenses/gpl-2.0.txt
#

from flask import Blueprint, request, current_app
from api.lib.common import (make_ok, make_error, make_bad_request, document_using)

from api.lib.utils import accepted_url, first_init_admin_access
from apimethods.utils import is_json_boolean, is_json_true, is_valid_ipv4
from apimethods.system.network import dns_resolution, get_interfaces, get_interface, set_interfaces_roles, \
    get_interface_traffic, get_traffic_stats, put_interface, get_fqdn_api
from uuid import UUID
from json import loads
from api.lib.auth import admin_permission, logged_permission

blueprint = Blueprint(__name__, __name__)
# TODO: get the default user from a configuration file


@blueprint.route('/<system_id>/network/interface', methods=['GET'])
@document_using('static/apidocs/system.html')
@logged_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}})
def get_system_network_interfaces(system_id):
    """
    Return a list of system network interfaces
    """
    (success, data) = get_interfaces(system_id)
    if not success:
        return make_error(data, 500)

    return make_ok(interfaces=data)


@blueprint.route('/<system_id>/network/interface', methods=['PUT'])
@document_using('static/apidocs/system.html')
@admin_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}, 'interfaces': str})
def set_system_interfaces_roles(system_id):
    try:
        interfaces = loads(request.args.get("interfaces", None))
    except ValueError:
        current_app.logger.error("network: Bad 'interfaces' parameter. Must be a correct url encoded string or bad json data")
        return make_bad_request("Bad parameter data")

    (success, data) = set_interfaces_roles(system_id, interfaces)
    if not success:
        return make_error(data, 500)

    return make_ok(jobid=data)


@blueprint.route('/<system_id>/network/interface/<iface>', methods=['GET'])
@document_using('static/apidocs/system.html')
@logged_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}, 'iface': str})
def get_system_network_interface(system_id, iface):
    (success, data) = get_interface(system_id, iface)
    if not success:
        return make_error(data, 500)

    return make_ok(interface=data)


@blueprint.route('/<system_id>/network/interface/<iface>', methods=['PUT'])
@document_using('static/apidocs/system.html')
@admin_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}, 'iface': str, 'promisc': str})
def put_system_network_interface(system_id, iface):
    promisc = request.args.get("promisc")
    if promisc is not None:
        if not is_json_boolean(promisc):
            current_app.logger.error("network: put_system_network_interface error: Bad param 'promisc='%s'" % promisc)
            return make_bad_request("Bad param 'promisc=%s'" % promisc)
    else:
        current_app.logger.error("network: put_system_network_interface error: Missing parameter 'promisc'")
        return make_bad_request("Missing parameters")

    (success, msg) = put_interface(system_id, iface, is_json_true(promisc))
    if not success:
        current_app.logger.error("network: put_system_network_interface error: " + str(msg))
        return make_error(msg, 500)

    return make_ok()


@blueprint.route('/<system_id>/network/interface/<iface>/traffic', methods=['GET'])
@document_using('static/apidocs/system/network.html')
@admin_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}, 'iface': str, 'timeout': str})
def get_system_network_interface_traffic(system_id, iface):
    timeout = request.args.get('timeout', None)
    if timeout is None or timeout == '':
        timeout = 10
    (success, data) = get_interface_traffic(system_id, iface, timeout)
    if not success:
        current_app.logger.error("newtork: get_system_network_interface_traffic error: " + str(data))
        return make_error(data, 500)

    return make_ok(stats=data)


@blueprint.route('/<system_id>/network/traffic_stats', methods=['GET'])
@document_using('static/apidocs/system/network.html')
@admin_permission.require(http_exception=403)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}})
def get_system_network_traffic_stats(system_id):
    (success, data) = get_traffic_stats(system_id)
    if not success:
        current_app.logger.error("network: get_system_network_traffic_stats error: " + str(data))
        return make_error("Error getting iface list", 500)

    return make_ok(stats=data)


@blueprint.route('/<system_id>/network/resolve', methods=['GET'])
@document_using('static/apidocs/system/network.html')
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}})
def get_system_network_resolve(system_id):
    if not first_init_admin_access():
        return make_error('Request forbidden', 403)

    (success, data) = dns_resolution(system_id)
    if not success:
        current_app.logger.error("network: get_system_network_resolve error: " + str(data))
        return make_error(data, 500)

    return make_ok(dns_resolution=data)


@blueprint.route('/<system_id>/network/fqdn', methods=['POST'])
@document_using('static/apidocs/system/network.html')
@logged_permission.require(http_exception=401)
@accepted_url({'system_id': {'type': UUID, 'values': ['local']}, 'host_ip': str})
def get_host_fqdn(system_id):
    host_ip = request.form.get('host_ip')
    if not is_valid_ipv4(host_ip):
        return make_error("Invalid host IP address", 500)

    return make_ok(fqdn=get_fqdn_api(system_id, host_ip))
